// @ts-ioc/core/lib/core/factories/ArgsIterator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var utils_1 = require("../../utils");
/**
 * args iterator.s
 *
 * @export
 * @class ArgsIterator
 */
var ArgsIterator = /** @class */ (function () {
    function ArgsIterator(args) {
        this.args = args;
        this.idx = -1;
        this.metadata = null;
    }
    ArgsIterator.prototype.isCompeted = function () {
        return this.idx >= this.args.length;
    };
    ArgsIterator.prototype.end = function () {
        this.idx = this.args.length;
    };
    ArgsIterator.prototype.next = function (express) {
        this.idx++;
        if (this.isCompeted()) {
            return null;
        }
        var arg = this.args[this.idx];
        if (express.isMetadata && express.isMetadata(arg)) {
            this.metadata = utils_1.lang.assign(this.metadata || {}, arg);
            this.end();
        }
        else if (express.match(arg)) {
            this.metadata = this.metadata || {};
            express.setMetadata(this.metadata, arg);
        }
        else if (utils_1.isMetadataObject(arg)) { // when match failed then check is base metadata.
            this.metadata = utils_1.lang.assign(this.metadata || {}, arg);
            this.end();
        }
        else {
            this.end();
        }
    };
    ArgsIterator.prototype.getArgs = function () {
        return this.args;
    };
    ArgsIterator.prototype.getMetadata = function () {
        return this.metadata;
    };
    ArgsIterator.getClassAnnations = function () {
        return { "name": "ArgsIterator", "params": { "constructor": ["args"], "isCompeted": [], "end": [], "next": ["express"], "getArgs": [], "getMetadata": [] } };
    };
    return ArgsIterator;
}());
exports.ArgsIterator = ArgsIterator;
