// @ts-ioc/aop/lib/joinpoints/Joinpoint.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var IJoinpoint_1 = require("./IJoinpoint");
var core_1 = require("@ts-ioc/core");
var NonePointcut_1 = require("../decorators/NonePointcut");
/**
 * Join point data.
 *
 * @export
 * @class Joinpoint
 * @implements {IJoinpoint}
 */
var Joinpoint = /** @class */ (function () {
    function Joinpoint(options) {
        this.provJoinpoint = options.provJoinpoint;
        this.name = options.name;
        this.fullName = options.fullName;
        this.params = options.params || [];
        this.args = options.args;
        this.returning = options.returning;
        this.throwing = options.throwing;
        this.state = options.state;
        this.advicer = options.advicer;
        this.annotations = options.annotations;
        this.target = options.target;
        this.targetType = options.targetType;
    }
    Joinpoint.getClassAnnations = function () {
        return { "name": "Joinpoint", "params": { "constructor": ["options"] } };
    };
    Joinpoint = tslib_1.__decorate([
        core_1.Injectable(IJoinpoint_1.JoinpointToken),
        NonePointcut_1.NonePointcut(),
        tslib_1.__metadata("design:paramtypes", [Object])
    ], Joinpoint);
    return Joinpoint;
}());
exports.Joinpoint = Joinpoint;
