// @ts-ioc/core/es2015/components/NullComponent.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * null component.
 *
 * @export
 * @class NullComponent
 * @implements {IComponent}
 */
class NullComponent {
    isEmpty() {
        return true;
    }
    add(action) {
        return this;
    }
    remove(action) {
        return this;
    }
    find(express, mode) {
        return exports.NullNode;
    }
    filter(express, mode) {
        return [];
    }
    each(express, mode) {
    }
    trans(express) {
    }
    transAfter(express) {
    }
    routeUp(express) {
    }
    equals(node) {
        return node === exports.NullNode;
    }
    empty() {
        return exports.NullNode;
    }
    static getClassAnnations() {
        return { "name": "NullComponent", "params": { "isEmpty": [], "add": ["action"], "remove": ["action"], "find": ["express", "mode"], "filter": ["express", "mode"], "each": ["express", "mode"], "trans": ["express"], "transAfter": ["express"], "routeUp": ["express"], "equals": ["node"], "empty": [] } };
    }
}
exports.NullComponent = NullComponent;
/**
 * Null node
 */
exports.NullNode = new NullComponent();
