// @ts-ioc/core/es2015/providers/Provider.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const utils_1 = require("../utils");
/**
 *  provider, to dynamic resovle instance of params in run time.
 *
 * @export
 * @class Provider
 */
class Provider {
    constructor(type, value) {
        this.type = type;
        this.value = value;
    }
    /**
     * resolve provider value.
     *
     * @template T
     * @param {IContainer} container
     * @param {ProviderTypes[]} providers
     * @returns {T}
     * @memberof Provider
     */
    resolve(container, ...providers) {
        if (utils_1.isUndefined(this.value)) {
            return container.has(this.type) ? container.resolve(this.type, ...providers) : null;
        }
        else {
            return this.value;
        }
    }
    /**
     * create provider.
     *
     * @static
     * @param {Token<any>} type
     * @param {(any)} value
     * @returns Provider
     * @memberof Provider
     */
    static create(type, value) {
        return new Provider(type, value);
    }
    /**
     * create extends provider.
     *
     * @static
     * @param {Token<any>} token
     * @param {(any)} value
     * @param {Express2<any, ExtendsProvider, void>} [extendsTarget]
     * @returns {ExtendsProvider}
     * @memberof Provider
     */
    static createExtends(token, value, extendsTarget) {
        return new ExtendsProvider(token, value, extendsTarget);
    }
    /**
     * create invoked provider.
     *
     * @static
     * @param {Token<any>} token
     * @param {string} method
     * @param {(any)} [value]
     * @returns {InvokeProvider}
     * @memberof Provider
     */
    static createInvoke(token, method, value) {
        return new InvokeProvider(token, method, value);
    }
    /**
     * create param provider.
     *
     * @static
     * @param {Token<any>} token
     * @param {(any)} value
     * @param {number} [index]
     * @param {string} [method]
     * @returns {ParamProvider}
     * @memberof Provider
     */
    static createParam(token, value, index, method) {
        return new ParamProvider(token, value, index, method);
    }
    static getClassAnnations() {
        return { "name": "Provider", "params": { "constructor": ["type", "value"], "resolve": ["container", "providers"], "create": ["type", "value"], "createExtends": ["token", "value", "extendsTarget"], "createInvoke": ["token", "method", "value"], "createParam": ["token", "value", "index", "method"] } };
    }
}
exports.Provider = Provider;
/**
 * InvokeProvider
 *
 * @export
 * @class InvokeProvider
 * @extends {Provider}
 */
class InvokeProvider extends Provider {
    constructor(type, method, value) {
        super(type, value);
        this.method = method;
    }
    resolve(container, ...providers) {
        if (this.method) {
            return container.syncInvoke(this.type, this.method, ...providers);
        }
        return super.resolve(container, ...providers);
    }
    static getClassAnnations() {
        return { "name": "InvokeProvider", "params": { "constructor": ["type", "method", "value"], "resolve": ["container", "providers"] } };
    }
}
exports.InvokeProvider = InvokeProvider;
/**
 * param provider.
 *
 * @export
 * @interface ParamProvider
 */
class ParamProvider extends InvokeProvider {
    constructor(token, value, index, method) {
        super(token, method, value);
        this.index = index;
    }
    /**
     * resolve param
     *
     * @template T
     * @param {IContainer} container
     * @param {...ProviderTypes[]} providers
     * @returns {T}
     * @memberof ParamProvider
     */
    resolve(container, ...providers) {
        return super.resolve(container, ...providers);
    }
    static getClassAnnations() {
        return { "name": "ParamProvider", "params": { "constructor": ["token", "value", "index", "method"], "resolve": ["container", "providers"] } };
    }
}
exports.ParamProvider = ParamProvider;
/**
 * Provider enable exntends target with provider in dynamic.
 *
 * @export
 * @class ExtendsProvider
 * @extends {Provider}
 */
class ExtendsProvider extends Provider {
    constructor(token, value, extendsTarget) {
        super(token, value);
        this.extendsTarget = extendsTarget;
    }
    resolve(container, ...providers) {
        return super.resolve(container, ...providers);
    }
    extends(target) {
        if (utils_1.isObject(target) && utils_1.isFunction(this.extendsTarget)) {
            this.extendsTarget(target, this);
        }
    }
    static getClassAnnations() {
        return { "name": "ExtendsProvider", "params": { "constructor": ["token", "value", "extendsTarget"], "resolve": ["container", "providers"], "extends": ["target"] } };
    }
}
exports.ExtendsProvider = ExtendsProvider;
