// @ts-ioc/core/es2015/injectors/ModuleValidate.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const IModuleValidate_1 = require("./IModuleValidate");
const utils_1 = require("../utils");
const core_1 = require("../core");
/**
 * base module validate.
 *
 * @export
 * @abstract
 * @class BaseModuelValidate
 * @implements {IModuleValidate}
 */
class ModuelValidate {
    constructor() {
    }
    valid(type) {
        if (!utils_1.isClass(type)) {
            return false;
        }
        let decorator = this.getDecorator();
        if (utils_1.isString(decorator)) {
            return core_1.hasOwnClassMetadata(decorator, type);
        }
        else if (utils_1.isArray(decorator)) {
            return decorator.some(d => core_1.hasOwnClassMetadata(d, type));
        }
        return true;
    }
    getDecorator() {
        return null;
    }
    static getClassAnnations() {
        return { "name": "ModuelValidate", "params": { "constructor": [], "valid": ["type"], "getDecorator": [] } };
    }
}
exports.ModuelValidate = ModuelValidate;
/**
 * IocExt module validate token.
 */
exports.IocExtModuleValidateToken = new IModuleValidate_1.InjectModuleValidateToken(core_1.IocExt.toString());
/**
 * IocExt module validate.
 *
 * @export
 * @class IocExtModuleValidate
 * @extends {ModuelValidate}
 * @implements {IModuleValidate}
 */
class IocExtModuleValidate extends ModuelValidate {
    getDecorator() {
        return core_1.IocExt.toString();
    }
    static getClassAnnations() {
        return { "name": "IocExtModuleValidate", "params": { "getDecorator": [] } };
    }
}
exports.IocExtModuleValidate = IocExtModuleValidate;
