// @ts-ioc/core/lib/components/NullComponent.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * null component.
 *
 * @export
 * @class NullComponent
 * @implements {IComponent}
 */
var NullComponent = /** @class */ (function () {
    function NullComponent() {
    }
    NullComponent.prototype.isEmpty = function () {
        return true;
    };
    NullComponent.prototype.add = function (action) {
        return this;
    };
    NullComponent.prototype.remove = function (action) {
        return this;
    };
    NullComponent.prototype.find = function (express, mode) {
        return exports.NullNode;
    };
    NullComponent.prototype.filter = function (express, mode) {
        return [];
    };
    NullComponent.prototype.each = function (express, mode) {
    };
    NullComponent.prototype.trans = function (express) {
    };
    NullComponent.prototype.transAfter = function (express) {
    };
    NullComponent.prototype.routeUp = function (express) {
    };
    NullComponent.prototype.equals = function (node) {
        return node === exports.NullNode;
    };
    NullComponent.prototype.empty = function () {
        return exports.NullNode;
    };
    NullComponent.getClassAnnations = function () {
        return { "name": "NullComponent", "params": { "isEmpty": [], "add": ["action"], "remove": ["action"], "find": ["express", "mode"], "filter": ["express", "mode"], "each": ["express", "mode"], "trans": ["express"], "transAfter": ["express"], "routeUp": ["express"], "equals": ["node"], "empty": [] } };
    };
    return NullComponent;
}());
exports.NullComponent = NullComponent;
/**
 * Null node
 */
exports.NullNode = new NullComponent();
