// @ts-ioc/core/lib/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * reference target level.
 *
 * @export
 * @enum {number}
 */
var RefTagLevel;
(function (RefTagLevel) {
    /**
     * ref taget self only
     */
    RefTagLevel[RefTagLevel["self"] = 1] = "self";
    /**
     * ref taget provider.
     */
    RefTagLevel[RefTagLevel["providers"] = 2] = "providers";
    /**
     * self provider
     */
    RefTagLevel[RefTagLevel["selfProviders"] = 3] = "selfProviders";
    /**
     * ref target class chain.
     */
    RefTagLevel[RefTagLevel["chain"] = 4] = "chain";
    /**
     * self chain.
     */
    RefTagLevel[RefTagLevel["selfChain"] = 5] = "selfChain";
    /**
     * chain providers.
     */
    RefTagLevel[RefTagLevel["chainProviders"] = 6] = "chainProviders";
    /**
     * ref all.
     */
    RefTagLevel[RefTagLevel["all"] = 7] = "all";
})(RefTagLevel = exports.RefTagLevel || (exports.RefTagLevel = {}));
/**
 * State of type in ioc.
 *
 * @export
 * @enum {number}
 */
var IocState;
(function (IocState) {
    IocState["design"] = "design";
    IocState["runtime"] = "runtime";
})(IocState = exports.IocState || (exports.IocState = {}));
/**
 * iterate way.
 *
 * @export
 * @enum {number}
 */
var Mode;
(function (Mode) {
    /**
     * route up. iterate in parents.
     */
    Mode[Mode["route"] = 1] = "route";
    /**
     * iterate in children.
     */
    Mode[Mode["children"] = 2] = "children";
    /**
     * iterate as tree map. node first
     */
    Mode[Mode["traverse"] = 3] = "traverse";
    /**
     * iterate as tree map. node last
     */
    Mode[Mode["traverseLast"] = 4] = "traverseLast";
})(Mode = exports.Mode || (exports.Mode = {}));
