// @ts-ioc/core/lib/injectors/ModuleValidate.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var IModuleValidate_1 = require("./IModuleValidate");
var utils_1 = require("../utils");
var core_1 = require("../core");
/**
 * base module validate.
 *
 * @export
 * @abstract
 * @class BaseModuelValidate
 * @implements {IModuleValidate}
 */
var ModuelValidate = /** @class */ (function () {
    function ModuelValidate() {
    }
    ModuelValidate.prototype.valid = function (type) {
        if (!utils_1.isClass(type)) {
            return false;
        }
        var decorator = this.getDecorator();
        if (utils_1.isString(decorator)) {
            return core_1.hasOwnClassMetadata(decorator, type);
        }
        else if (utils_1.isArray(decorator)) {
            return decorator.some(function (d) { return core_1.hasOwnClassMetadata(d, type); });
        }
        return true;
    };
    ModuelValidate.prototype.getDecorator = function () {
        return null;
    };
    ModuelValidate.getClassAnnations = function () {
        return { "name": "ModuelValidate", "params": { "constructor": [], "valid": ["type"], "getDecorator": [] } };
    };
    return ModuelValidate;
}());
exports.ModuelValidate = ModuelValidate;
/**
 * IocExt module validate token.
 */
exports.IocExtModuleValidateToken = new IModuleValidate_1.InjectModuleValidateToken(core_1.IocExt.toString());
/**
 * IocExt module validate.
 *
 * @export
 * @class IocExtModuleValidate
 * @extends {ModuelValidate}
 * @implements {IModuleValidate}
 */
var IocExtModuleValidate = /** @class */ (function (_super) {
    tslib_1.__extends(IocExtModuleValidate, _super);
    function IocExtModuleValidate() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    IocExtModuleValidate.prototype.getDecorator = function () {
        return core_1.IocExt.toString();
    };
    IocExtModuleValidate.getClassAnnations = function () {
        return { "name": "IocExtModuleValidate", "params": { "getDecorator": [] } };
    };
    return IocExtModuleValidate;
}(ModuelValidate));
exports.IocExtModuleValidate = IocExtModuleValidate;
