// @ts-ioc/aop/es2015/joinpoints/Joinpoint.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const IJoinpoint_1 = require("./IJoinpoint");
const core_1 = require("@ts-ioc/core");
const NonePointcut_1 = require("../decorators/NonePointcut");
/**
 * Join point data.
 *
 * @export
 * @class Joinpoint
 * @implements {IJoinpoint}
 */
let Joinpoint = class Joinpoint {
    constructor(options) {
        this.provJoinpoint = options.provJoinpoint;
        this.name = options.name;
        this.fullName = options.fullName;
        this.params = options.params || [];
        this.args = options.args;
        this.returning = options.returning;
        this.throwing = options.throwing;
        this.state = options.state;
        this.advicer = options.advicer;
        this.annotations = options.annotations;
        this.target = options.target;
        this.targetType = options.targetType;
    }
    static getClassAnnations() {
        return { "name": "Joinpoint", "params": { "constructor": ["options"] } };
    }
};
Joinpoint = tslib_1.__decorate([
    core_1.Injectable(IJoinpoint_1.JoinpointToken),
    NonePointcut_1.NonePointcut(),
    tslib_1.__metadata("design:paramtypes", [Object])
], Joinpoint);
exports.Joinpoint = Joinpoint;
