// @ts-ioc/core/es2015/resolves/ResolverChain.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const IContainer_1 = require("../IContainer");
const InjectToken_1 = require("../InjectToken");
const providers_1 = require("../providers");
const utils_1 = require("../utils");
/**
 *  resolver chain token.
 */
exports.ResolverChainToken = new InjectToken_1.InjectToken('di_ResolverChain');
/**
 * resover chain.
 *
 * resove by setp.
 *
 * @export
 * @class ResolverChain
 * @implements {IResolver}
 */
class ResolverChain {
    constructor(container) {
        this.container = container;
        this.resolvers = [];
    }
    get size() {
        return this.resolvers.length + 1;
    }
    /**
     * reigister next resolver.
     *
     * @param {IResolver} resolver
     * @memberof ResolverChain
     */
    next(resolver) {
        if (!this.hasResolver(resolver)) {
            this.resolvers.push(resolver);
        }
    }
    /**
     * has resolver or not.
     *
     * @param {IResolver} resolver
     * @returns
     * @memberof ResolverChain
     */
    hasResolver(resolver) {
        if (resolver === this.container) {
            return true;
        }
        return this.resolvers.indexOf(resolver) >= 0;
    }
    /**
     * resove token via registered resolver chain.
     *
     * @template T
     * @param {SymbolType<T>} token
     * @param {...ParamProviders[]} providers
     * @returns {T}
     * @memberof ResolverChain
     */
    resolve(token, resway, ...providers) {
        let key = this.container.getTokenKey(token, utils_1.isString(resway) ? resway : null);
        let way;
        if (utils_1.isNumber(resway)) {
            way = resway;
        }
        else {
            if (resway) {
                providers.unshift(resway);
            }
            way = IContainer_1.ResoveWay.all;
        }
        let providerMap;
        if (providers.length) {
            if (providers.length === 1 && providers_1.isProviderMap(providers[0])) {
                providerMap = providers[0];
            }
            else {
                providerMap = this.container.getProviderParser().parse(...providers);
            }
        }
        if (providerMap && providerMap.has(token)) {
            return providerMap.resolve(token, providerMap);
        }
        if ((way & IContainer_1.ResoveWay.current) && this.container.hasRegister(key)) {
            return this.container.resolveValue(key, providerMap);
        }
        if ((way & IContainer_1.ResoveWay.traverse)) {
            let resolver = this.resolvers.find(r => r.has(key, IContainer_1.ResoveWay.nodes));
            if (resolver) {
                return resolver.resolve(key, IContainer_1.ResoveWay.nodes, providerMap);
            }
        }
        if (this.container.parent && (way & IContainer_1.ResoveWay.bubble)) {
            return this.container.parent.resolve(key, resway, providerMap);
        }
        return null;
    }
    /**
     * unregister token in registered resolver chain.
     *
     * @template T
     * @param {SymbolType<T>} token
     * @memberof ResolverChain
     */
    unregister(token, resway) {
        resway = resway || IContainer_1.ResoveWay.all;
        let tokenKey = this.container.getTokenKey(token);
        if (resway & IContainer_1.ResoveWay.current) {
            this.container.unregisterValue(tokenKey);
        }
        if (resway & IContainer_1.ResoveWay.traverse) {
            this.resolvers.forEach((r) => {
                if (utils_1.isFunction(r.unregister)) {
                    r.unregister(tokenKey, IContainer_1.ResoveWay.nodes);
                }
            });
        }
        if ((resway & IContainer_1.ResoveWay.bubble) && this.container.parent) {
            this.container.parent.unregister(token, resway);
        }
        return this;
    }
    /**
     * get token implements class in the registered resolver chain.
     *
     * @template T
     * @param {Token<T>} token
     * @returns {Type<T>}
     * @memberof ResolverChain
     */
    getTokenImpl(token, resway) {
        if (utils_1.isClass(token)) {
            return token;
        }
        resway = resway || IContainer_1.ResoveWay.nodes;
        let tokenKey = this.container.getTokenKey(token);
        let provider;
        if (resway & IContainer_1.ResoveWay.current) {
            provider = this.container.getTokenProvider(tokenKey);
        }
        if (!provider && (resway & IContainer_1.ResoveWay.traverse)) {
            this.resolvers.some((r) => {
                if (!utils_1.isFunction(r.getTokenImpl)) {
                    return false;
                }
                provider = r.getTokenImpl(tokenKey, IContainer_1.ResoveWay.nodes);
                return !!provider;
            });
        }
        if (!provider && (resway & IContainer_1.ResoveWay.bubble) && this.container.parent) {
            provider = this.container.parent.getTokenImpl(token, resway);
        }
        return provider || null;
    }
    /**
     * has token or not in the registered resolver chain.
     *
     * @template T
     * @param {SymbolType<T>} token
     * @returns {boolean}
     * @memberof ResolverChain
     */
    has(token, aliasOrway) {
        let key = this.container.getTokenKey(token, utils_1.isString(aliasOrway) ? aliasOrway : null);
        let resway = utils_1.isNumber(aliasOrway) ? aliasOrway : IContainer_1.ResoveWay.all;
        if ((resway & IContainer_1.ResoveWay.current) && this.container.hasRegister(key)) {
            return true;
        }
        if ((resway & IContainer_1.ResoveWay.traverse) && this.resolvers.some(r => r.has(key, IContainer_1.ResoveWay.nodes))) {
            return true;
        }
        if ((resway & IContainer_1.ResoveWay.bubble) && this.container.parent) {
            return this.container.parent.has(token, resway);
        }
        return false;
    }
    /**
     * resolver chain to array.
     *
     * @returns {IResolver[]}
     * @memberof ResolverChain
     */
    toArray(resway = IContainer_1.ResoveWay.all) {
        if (resway & IContainer_1.ResoveWay.nodes) {
            return [this.container].concat(this.resolvers);
        }
        else if (resway & IContainer_1.ResoveWay.current) {
            return [this.container];
        }
        else if (resway & IContainer_1.ResoveWay.traverse) {
            return this.resolvers;
        }
        return [];
    }
    forEach(callbackfn) {
        if (this.container.forEach(callbackfn) === false) {
            return false;
        }
        return !this.resolvers.some((r) => {
            if (utils_1.isFunction(r.forEach)) {
                return r.forEach(callbackfn) === false;
            }
            return false;
        });
    }
    /**
     * iterator all resolvers.
     *
     * @param {(tk: Token<any>, fac: InstanceFactory<any>, resolvor?: IResolver) => void} callbackfn
     * @memberof ResolverChain
     */
    iterator(callbackfn, resway = IContainer_1.ResoveWay.all) {
        if (resway & IContainer_1.ResoveWay.current) {
            if (this.container.forEach(callbackfn) === false) {
                return false;
            }
        }
        if (resway & IContainer_1.ResoveWay.traverse) {
            if (this.resolvers.some((r) => {
                if (utils_1.isFunction(r.forEach)) {
                    return r.forEach(callbackfn) === false;
                }
                return false;
            })) {
                return false;
            }
        }
        if (this.container.parent && (resway & IContainer_1.ResoveWay.bubble)) {
            return this.container.parent.iterator(callbackfn, resway);
        }
    }
    static getClassAnnations() {
        return { "name": "ResolverChain", "params": { "constructor": ["container"], "next": ["resolver"], "hasResolver": ["resolver"], "resolve": ["token", "resway", "providers"], "unregister": ["token", "resway"], "getTokenImpl": ["token", "resway"], "has": ["token", "aliasOrway"], "toArray": ["resway"], "forEach": ["callbackfn"], "iterator": ["callbackfn", "resway"] } };
    }
}
exports.ResolverChain = ResolverChain;
