// @ts-ioc/core/es2015/Registration.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const utils_1 = require("./utils");
/**
 * is registration class or not.
 *
 * @export
 * @param {*} target
 * @returns
 */
function isRegistrationClass(target) {
    if (utils_1.isClass(target)) {
        return target.isIocRegClass === true;
    }
    return false;
}
exports.isRegistrationClass = isRegistrationClass;
/**
 * inject token.
 * @export
 * @class Registration
 * @template T
 */
class Registration {
    /**
     * Creates an instance of Registration.
     * @param {(Token<T> | Token<any>)} provideType
     * @param {string} desc
     * @memberof Registration
     */
    constructor(provideType, desc) {
        this.type = '';
        this.init(provideType, desc);
    }
    init(provideType, desc) {
        if (provideType instanceof Registration) {
            if (desc) {
                this.classType = provideType.toString();
                this.desc = desc;
            }
            else {
                this.classType = provideType.getProvide();
                this.desc = provideType.getDesc();
            }
        }
        else {
            this.classType = provideType;
            this.desc = desc;
        }
    }
    /**
     * get provide.
     *
     * @returns {SymbolType<any>}
     * @memberof Registration
     */
    getProvide() {
        return this.classType;
    }
    /**
     * get class.
     *
     * @returns
     * @memberof Registration
     */
    getClass() {
        if (utils_1.isClass(this.classType)) {
            return this.classType;
        }
        return null;
    }
    /**
     * get desc.
     *
     * @returns
     * @memberof Registration
     */
    getDesc() {
        return this.desc;
    }
    /**
     * to string.
     *
     * @returns {string}
     * @memberof Registration
     */
    toString() {
        return this.format(this);
    }
    format(reg) {
        if (reg instanceof Registration) {
            let name = '';
            if (utils_1.isFunction(reg.classType)) {
                name = `{${utils_1.lang.getClassName(reg.classType)}}`;
            }
            else if (reg.classType) {
                name = reg.classType.toString();
            }
            return [reg.type, name, reg.desc].filter(n => n).join('_');
        }
        else if (utils_1.isFunction(reg)) {
            return `{${utils_1.lang.getClassName(reg)}}`;
        }
        else if (reg) {
            return reg.toString();
        }
        return '';
    }
    static getClassAnnations() {
        return { "name": "Registration", "params": { "constructor": ["provideType", "desc"], "init": ["provideType", "desc"], "getProvide": [], "getClass": [], "getDesc": [], "toString": [], "format": ["reg"] } };
    }
}
Registration.isIocRegClass = true;
exports.Registration = Registration;
