// @ts-ioc/core/es2015/core/actions/ActionComposite.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const NullAction_1 = require("./NullAction");
const components_1 = require("../../components");
/**
 * action composite
 *
 * @export
 * @class ActionComposite
 * @extends {GComposite<ActionComponent>}
 * @implements {ActionComponent}
 */
class ActionComposite extends components_1.GComposite {
    constructor(name) {
        super(name);
        this.children = [];
    }
    insert(node, index) {
        node.parent = this;
        if (index < 0) {
            index = 0;
        }
        else if (index >= this.children.length) {
            index = this.children.length - 1;
        }
        this.children.splice(index, 0, node);
        return this;
    }
    execute(container, data, name) {
        if (name) {
            this.find(it => it.name === name)
                .execute(container, data);
        }
        else {
            this.trans(action => {
                if (action instanceof ActionComposite) {
                    action.working(container, data);
                }
            });
        }
    }
    empty() {
        return NullAction_1.NullAction;
    }
    working(container, data) {
        // do nothing.
    }
    static getClassAnnations() {
        return { "name": "ActionComposite", "params": { "constructor": ["name"], "insert": ["node", "index"], "execute": ["container", "data", "name"], "empty": [], "working": ["container", "data"] } };
    }
}
exports.ActionComposite = ActionComposite;
