// @ts-ioc/aop/lib/access/ProxyMethod.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var core_1 = require("@ts-ioc/core");
var joinpoints_1 = require("../joinpoints");
var joinpoints_2 = require("../joinpoints");
var IAdvisor_1 = require("../IAdvisor");
var IProxyMethod_1 = require("./IProxyMethod");
var IAdvisorChainFactory_1 = require("./IAdvisorChainFactory");
var NonePointcut_1 = require("../decorators/NonePointcut");
/**
 * Proxy method.
 *
 * @export
 * @class ProxyMethod
 * @implements {IProxyMethod}
 */
var ProxyMethod = /** @class */ (function () {
    function ProxyMethod(container) {
        this.container = container;
    }
    Object.defineProperty(ProxyMethod.prototype, "advisor", {
        get: function () {
            if (!this._advisor) {
                this._advisor = this.container.get(IAdvisor_1.AdvisorToken);
            }
            return this._advisor;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ProxyMethod.prototype, "liefScope", {
        get: function () {
            if (!this._liefScope) {
                this._liefScope = this.container.getLifeScope();
            }
            return this._liefScope;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * proceed the proxy method.
     *
     * @param {*} target
     * @param {Type<any>} targetType
     * @param {IPointcut} pointcut
     * @param {Joinpoint} [provJoinpoint]
     * @memberof ProxyMethod
     */
    ProxyMethod.prototype.proceed = function (target, targetType, pointcut, provJoinpoint) {
        var aspectMgr = this.advisor;
        var fullName = pointcut.fullName;
        var methodName = pointcut.name;
        var advices = aspectMgr.getAdvices(fullName);
        if (advices && pointcut) {
            if (pointcut.descriptor && (pointcut.descriptor.get || pointcut.descriptor.set)) {
                if (pointcut.descriptor.get) {
                    var getMethod = pointcut.descriptor.get.bind(target);
                    pointcut.descriptor.get = this.proxy(getMethod, advices, target, targetType, pointcut, provJoinpoint);
                }
                if (pointcut.descriptor.set) {
                    var setMethod = pointcut.descriptor.set.bind(target);
                    pointcut.descriptor.set = this.proxy(setMethod, advices, target, targetType, pointcut, provJoinpoint);
                }
                Object.defineProperty(target, methodName, pointcut.descriptor);
            }
            else if (core_1.isFunction(target[methodName])) {
                var propertyMethod = target[methodName].bind(target);
                target[methodName] = this.proxy(propertyMethod, advices, target, targetType, pointcut, provJoinpoint);
            }
        }
    };
    ProxyMethod.prototype.proxy = function (propertyMethod, advices, target, targetType, pointcut, provJoinpoint) {
        var _this = this;
        var fullName = pointcut.fullName;
        var methodName = pointcut.name;
        var liefScope = this.liefScope;
        var container = this.container;
        return function () {
            var args = [];
            for (var _i = 0; _i < arguments.length; _i++) {
                args[_i] = arguments[_i];
            }
            var joinPoint = _this.container.resolve(joinpoints_2.Joinpoint, core_1.Provider.create('options', {
                name: methodName,
                fullName: fullName,
                provJoinpoint: provJoinpoint,
                annotations: provJoinpoint ? null : liefScope.getMethodMetadatas(targetType, methodName),
                params: liefScope.getMethodParameters(targetType, target, methodName),
                args: args,
                target: target,
                targetType: targetType
            }));
            var adChain = container.resolve(IAdvisorChainFactory_1.AdvisorChainFactoryToken, { container: container, advisor: _this.advisor, advices: advices });
            adChain.invoaction(joinPoint, joinpoints_1.JoinpointState.Before);
            adChain.invoaction(joinPoint, joinpoints_1.JoinpointState.Pointcut);
            var val, exeErr;
            try {
                val = propertyMethod.apply(void 0, joinPoint.args);
            }
            catch (err) {
                exeErr = err;
            }
            adChain.invoaction(joinPoint, joinpoints_1.JoinpointState.After, val);
            if (exeErr) {
                adChain.invoaction(joinPoint, joinpoints_1.JoinpointState.AfterThrowing, exeErr);
            }
            else {
                adChain.invoaction(joinPoint, joinpoints_1.JoinpointState.AfterReturning, val);
                return joinPoint.returning;
            }
        };
    };
    ProxyMethod.getClassAnnations = function () {
        return { "name": "ProxyMethod", "params": { "constructor": ["container"], "proceed": ["target", "targetType", "pointcut", "provJoinpoint"], "proxy": ["propertyMethod", "advices", "target", "targetType", "pointcut", "provJoinpoint"] } };
    };
    ProxyMethod = tslib_1.__decorate([
        NonePointcut_1.NonePointcut(),
        core_1.Singleton(IProxyMethod_1.ProxyMethodToken),
        tslib_1.__param(0, core_1.Inject(core_1.ContainerToken)),
        tslib_1.__metadata("design:paramtypes", [Object])
    ], ProxyMethod);
    return ProxyMethod;
}());
exports.ProxyMethod = ProxyMethod;
