// @ts-ioc/core/es2015/core/factories/ArgsIterator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const utils_1 = require("../../utils");
/**
 * args iterator.s
 *
 * @export
 * @class ArgsIterator
 */
class ArgsIterator {
    constructor(args) {
        this.args = args;
        this.idx = -1;
        this.metadata = null;
    }
    isCompeted() {
        return this.idx >= this.args.length;
    }
    end() {
        this.idx = this.args.length;
    }
    next(express) {
        this.idx++;
        if (this.isCompeted()) {
            return null;
        }
        let arg = this.args[this.idx];
        if (express.isMetadata && express.isMetadata(arg)) {
            this.metadata = utils_1.lang.assign(this.metadata || {}, arg);
            this.end();
        }
        else if (express.match(arg)) {
            this.metadata = this.metadata || {};
            express.setMetadata(this.metadata, arg);
        }
        else if (utils_1.isMetadataObject(arg)) { // when match failed then check is base metadata.
            this.metadata = utils_1.lang.assign(this.metadata || {}, arg);
            this.end();
        }
        else {
            this.end();
        }
    }
    getArgs() {
        return this.args;
    }
    getMetadata() {
        return this.metadata;
    }
    static getClassAnnations() {
        return { "name": "ArgsIterator", "params": { "constructor": ["args"], "isCompeted": [], "end": [], "next": ["express"], "getArgs": [], "getMetadata": [] } };
    }
}
exports.ArgsIterator = ArgsIterator;
